sdk/src/types/block/core/wrapper.rs

```diff
@@ -0,0 +1,348 @@
+        let mut unpacker = CounterUnpacker::new(SliceUnpacker::new(bytes.as_ref()));
+        let block = Self::unpack::<_, true>(&mut unpacker, visitor)?;
+
+        // When parsing the block is complete, there should not be any trailing bytes left that were not parsed.
+        if u8::unpack::<_, true>(&mut unpacker, &()).is_ok() {
+            return Err(UnpackError::Packable(Error::RemainingBytesAfterBlock));
+        }
+
+        Ok(block)
+    }
+
+    /// Checks whether the inner block is a [`BasicBlock`].
+    pub fn is_basic(&self) -> bool {
+        self.block.is_basic()
+    }
+
+    /// Gets the inner block as an actual [`BasicBlock`].
+    /// NOTE: Will panic if the inner block is not a [`BasicBlock`].
+    pub fn as_basic(&self) -> &BasicBlock {
+        self.block.as_basic()
+    }
+
+    /// Checks whether the inner block is a [`ValidationBlock`].
+    pub fn is_validation(&self) -> bool {
+        self.block.is_validation()
+    }
+
+    /// Gets the inner block as an actual [`ValidationBlock`].
+    /// NOTE: Will panic if the inner block is not a [`ValidationBlock`].
+    pub fn as_validation(&self) -> &ValidationBlock {
+        self.block.as_validation()
+    }
+}
+
+impl Packable for BlockWrapper {
+    type UnpackError = Error;
+    type UnpackVisitor = ProtocolParameters;
+
+    fn pack<P: Packer>(&self, packer: &mut P) -> Result<(), P::Error> {
+        self.pack_header(packer)?;
+        self.block.pack(packer)?;
+        self.signature.pack(packer)?;
+
+        Ok(())
+    }
+
+    fn unpack<U: Unpacker, const VERIFY: bool>(
+        unpacker: &mut U,
+        protocol_params: &Self::UnpackVisitor,
+    ) -> Result<Self, UnpackError<Self::UnpackError, U::Error>> {
+        let start_opt = unpacker.read_bytes();
+
+        let protocol_version = u8::unpack::<_, VERIFY>(unpacker, &()).coerce()?;
+
+        if VERIFY && protocol_version != protocol_params.version() {
+            return Err(UnpackError::Packable(Error::ProtocolVersionMismatch {
+                expected: protocol_params.version(),
+                actual: protocol_version,
+            }));
+        }
+
+        let network_id = u64::unpack::<_, VERIFY>(unpacker, &()).coerce()?;
+
+        if VERIFY && network_id != protocol_params.network_id() {
+            return Err(UnpackError::Packable(Error::NetworkIdMismatch {
+                expected: protocol_params.network_id(),
+                actual: network_id,
+            }));
+        }
+
+        let issuing_time = u64::unpack::<_, VERIFY>(unpacker, &()).coerce()?;
+
+        let slot_commitment_id = SlotCommitmentId::unpack::<_, VERIFY>(unpacker, &()).coerce()?;
+
+        let latest_finalized_slot = SlotIndex::unpack::<_, VERIFY>(unpacker, &()).coerce()?;
+
+        let issuer_id = IssuerId::unpack::<_, VERIFY>(unpacker, &()).coerce()?;
+
+        let block = Block::unpack::<_, VERIFY>(unpacker, protocol_params)?;
+
+        let signature = Signature::unpack::<_, VERIFY>(unpacker, &())?;
```

Comment:
personal preference: as for packing I would probably not put new-lines in between individual unpacks.